RFCs/0017-TestCase-FullyQualifiedName.md

```diff
@@ -1,49 +1,132 @@
+
+## Syntactic vs Semantic Test Location
+
+We are defining syntactic location as the location in the code/syntax where the method is defined that implements the test. On the other hand, semantic location is the location in the type hierarchy at which point the method becomes recognized as a valid test. 
+
+There can be a situation where a test is declared on an abstract base type (syntactic), but the test is only discovered in a derived class (semantic). The question is which type should the FQN point to? The base type, where the test code is located? Or the derived type, where the test is discovered. We will use the semantic location for the type because that is where the test adapter will find the test and it is where the test class would be instantiated. Also, there is a one to many relationship between the two. There can be many semantic test locations that each are implemented by the same base class method. If we were to standardize on using the syntactic location, we would risk having name collisions or fail to find the correct test case since there are multiple possible candidates.
+
+## Managed Tests that are Not Type/Method Based
+
+It is possible to create unit tests in managed code where tests are not necessarily based on types & methods. While this is somewhat uncommon, we need to be able to handle this situation gracefully. Some examples are gherkin-dotnet and custom discoverers in xUnit. In these examples, an externally parsed definition file may determine the tests. For these cases, we have defined the 'other' form of FQN. FQNs in this form will look similar to the following.
+
+    fqn://clr/o/{extension}/{other}/{url}/{segments}…
+
+The `'o'` indicates to the test window that we cannot interpret the contents of the FQN. The `{extension}` segment is to allow each adapter to add a custom string identifying the format. The rest of the URL is completely up to the interpretation of the test adapter. Type and method information will not be interpreted. Note that adapters using this format may lose hierarchy or code location benefits that are provided by the class/method FQN structure. There may be other ways to provide this information such as other properties on the test case. Those mechanisms are outside the scope of this document.
+
+## Legacy Compatibility
+
+We intend to implement a shared routine that can convert a legacy FQN into this standardized format. Throughout the TestWindow code, in all other locations, we would use the new standardized format. Any edge cases in interpretation can likely be addressed in the conversion routine. This does mean, however, that when handling test cases from adapters that do not support the format, we will need to allocate another property to save the original legacy FQN value. This is necessary so that when we call back into the adapter, we supply the original FQN rather than the new one. 
```

Comment:
What are the compat plans for older VS with new adapter versions (containing new FQN)

Reply:
We plan to have a run setting that the new version of VS adds. This will tell the adapter to use the standardized FQN implementation.

Reply:
Can you please update the RFC regarding the same?

1. So while discovery, adapter will create new FQN when runsettings has some particular value but old FQN when that value is not present?
2. And while execution, adapter will check if the FQN is old or new one. If old, convert it to new standard FQN and then run the tests?
2. What happens with vstest task? Does task flow will always go with old FQN flow OR user needs to manually update value in runsettings?

Reply:
I've updated the document to clarify the run setting.

> So while discovery, adapter will create new FQN when runsettings has some particular value but old FQN when that value is not present?

Yes, the adapter can change whether it uses the std FQN or not based on the run setting. Clarified in the doc.

> And while execution, adapter will check if the FQN is old or new one. If old, convert it to new standard FQN and then run the tests?

During execution, the adapter should only be given the FQN that it created when doing discovery. If discovery and execution are happening across different versions then some one-time conversion might be required.

> What happens with vstest task? Does task flow will always go with old FQN flow OR user needs to manually update value in runsettings?

The user will have the option to update the runsettings, but I would expect that we would update the vstest task to provide the runsetting as soon as we complete the support.

Reply:
>The user will have the option to update the runsettings

This value is not effecting any user experience. So, I don't think user should be aware of runsettings value which effects FQN.

>but I would expect that we would update the vstest task to provide the runsetting as soon as we complete the support.

We have already shipped TFS in which this update in task is not possible. Back compat will not be possible there.